### ISSUE TITLE
Missing hook to extend --dist choices when adding a new scheduler

### ISSUE THREAD

**Issue by vitaly-krugl** [pytest-dev/pytest-xdist#970]
Documentation says to use `pytest_xdist_make_scheduler` to add a new scheduler. However, option parsing fails because the choices for the `--dist` command line option are fixed by `xdist.plugin.pytest_addoption`.

```
$ pytest --basetemp ~/xdist-debug/basetemp/ -p xdist_schedulers.plugin -n 3 --dist mycoolnewscheduler
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: argument --dist: invalid choice: 'mycoolnewscheduler' (choose from 'each', 'load', 'loadscope', 'loadfile', 'loadgroup', 'worksteal', 'no')
```

**Comment by nicoddemus** [comment]
Hi @vitaly-krugl,

Indeed this is a problem that we overlooked.

From the top of my head I don't know how we could solve this, TBH.

**Comment by vitaly-krugl** [comment]
I ended up working around this limitation as follows:

1. Handle `pytest_addoption` hook and add a group/option for my own scheduler to the parser
2. Handle `pytest_cmdline_parse` with `hookwrapper=True` where I overwrite `config.option.dist` with the name of my own scheduler option dist option value.
3. Handle `pytest_xdist_make_scheduler` and instantiate my own scheduler if `config.option.dist` matches mine.

**Comment by vitaly-krugl** [comment]
@nicoddemus and @RonnyPfannschmidt, the workaround that I documented in my previous comment isn't necessarily a substitute for a well-documented and supported scheduler plugin mechanism. However, it demonstrates a working prototype for bootstrapping an add-on xdist scheduler into xdist and into the pytest command-line. It also suggests that an add-on scheduler may need to set its own options via pytest command line.

**Comment by amezin** [comment]
Maybe there should be a hook that returns a list of available schedulers (or a name => callable (factory) dictionary)?

**Comment by nicoddemus** [comment]
> the workaround that I documented in my previous comment isn't necessarily a substitute for a well-documented and supported scheduler plugin mechanism

Definitely, unfortunately (at least to me) there is no clear way forward yet on how to support this nicely.

**Comment by vitaly-krugl** [comment]
Hi @nicoddemus, here is a proposal that I think could work:

Presently, when xdist's plugin.pytest_configure would finds the default "no" in config.option.dist, it bypasses distributed execution altogether (wouldn't create DSession, etc.).

1. The proposed change is to not do the above. Instead, consider `config.option.dist` to be the name of the built-in scheduler and assume that other scheduler providers may have their own command line options for their own scheduler.
2. Modify xdist's handler of `pytest_xdist_make_scheduler` to return `None` if `config.option.dist == 'no'.  
3. Invoke the hook `pytest_xdist_make_scheduler`. If the hook doesn't return a scheduler, then - and only then - bypass distributed execution.
